src/IRM/IRMAdaptiveCurve.sol

```diff
@@ -0,0 +1,182 @@
+contract IRMAdaptiveCurve is IIRM {
+    /// @dev Unit for internal precision.
+    int256 internal constant WAD = 1e18;
+    /// @notice The utilization rate targeted by the model.
+    /// @dev In WAD units.
+    int256 public immutable TARGET_UTILIZATION;
+    /// @notice The initial interest rate at target utilization.
+    /// @dev In WAD per second units.
+    /// When the IRM is initialized for a vault this is the rate at target utilization that is assigned.
+    int256 public immutable INITIAL_RATE_AT_TARGET;
+    /// @notice The minimum interest rate at target utilization that the model can adjust to.
+    /// @dev In WAD per second units.
+    int256 public immutable MIN_RATE_AT_TARGET;
+    /// @notice The maximum interest rate at target utilization that the model can adjust to.
+    /// @dev In WAD per second units.
+    int256 public immutable MAX_RATE_AT_TARGET;
+    /// @notice The slope of interest rate line above the target. The line below the target has inverse slope.
+    /// @dev In WAD units.
+    int256 public immutable CURVE_STEEPNESS;
+    /// @notice The speed at which the rate at target is adjusted up or down.
+    /// @dev In WAD per second units.
+    /// For example, with `2e18 / 24 hours` the model will 2x `rateAtTarget` if the vault is fully utilized for a day.
+    int256 public immutable ADJUSTMENT_SPEED;
+
+    /// @notice Internal cached state of the interest rate model.
+    struct IRState {
+        /// @dev The current rate at target utilization.
+        uint208 rateAtTarget;
+        /// @dev The timestamp of the last update to the model.
+        uint48 lastUpdate;
+    }
+
+    /// @notice Get the cached state of a vault's irm.
+    mapping(address => IRState) internal irState;
+
+    /// @notice Deploy IRMAdaptiveCurve.
+    /// @param _TARGET_UTILIZATION The utilization rate targeted by the interest rate model.
+    /// @param _INITIAL_RATE_AT_TARGET The initial interest rate at target utilization.
+    /// @param _MIN_RATE_AT_TARGET The minimum interest rate at target utilization that the model can adjust to.
+    /// @param _MAX_RATE_AT_TARGET The maximum interest rate at target utilization that the model can adjust to.
+    /// @param _CURVE_STEEPNESS The slope of interest rate above target. The line below target has inverse slope.
+    /// @param _ADJUSTMENT_SPEED The speed at which the rate at target utilization is adjusted up or down.
+    constructor(
+        int256 _TARGET_UTILIZATION,
+        int256 _INITIAL_RATE_AT_TARGET,
+        int256 _MIN_RATE_AT_TARGET,
+        int256 _MAX_RATE_AT_TARGET,
+        int256 _CURVE_STEEPNESS,
+        int256 _ADJUSTMENT_SPEED
+    ) {
+        TARGET_UTILIZATION = _TARGET_UTILIZATION;
+        INITIAL_RATE_AT_TARGET = _INITIAL_RATE_AT_TARGET;
+        MIN_RATE_AT_TARGET = _MIN_RATE_AT_TARGET;
+        MAX_RATE_AT_TARGET = _MAX_RATE_AT_TARGET;
+        CURVE_STEEPNESS = _CURVE_STEEPNESS;
+        ADJUSTMENT_SPEED = _ADJUSTMENT_SPEED;
```

Comment:
to make sure the curve doesn't go negative I think it's enough to check that:
`CURVE_STEEPNESS * TARGET_UTILIZATION <= MIN_RATE_AT_TARGET` ie in `a` is positive in `ir(util) = slope*util + a`